osf/models/preprint_service.py

```diff
@@ -74,11 +74,25 @@ def deep_url(self):
 
     @property
     def url(self):
-        return '/{}/'.format(self._id)
+        if self.provider._id == 'osf' or hasattr(self.provider, 'domain'):
+            return '/{}/'.format(self._id)
+
+        return '/preprints/{}/{}/'.format(self.provider._id, self._id)
+
+    if settings.DEV_MODE:
+        def get_provider_domain(self):
+            domain_settings = settings.PREPRINT_PROVIDER_DOMAINS
+            return ''.join((domain_settings['prefix'], self.provider.domain, domain_settings['suffix']))
+    else:
+        def get_provider_domain(self):
+            return settings.PROTOCOL + self.provider.domain
 
     @property
     def absolute_url(self):
-        return urlparse.urljoin(settings.DOMAIN, self.url)
+        use_osf_domain = self.provider._id == 'osf' or not hasattr(self.provider, 'domain')
```

Comment:
Hruh, interesting. I assumed it would have the attribute in that case, but with a null value. Ok then.

(`hasattr` checks are something we used to do more often with schemaless data, where mongo sometimes literally did not have the field)